// Copyright (c) 2019 Boston Dynamics, Inc.  All rights reserved.
//
// Downloading, reproducing, distributing or otherwise using the SDK Software
// is subject to the terms and conditions of the Boston Dynamics Software
// Development Kit License (20191101-BDSDK-SL).

syntax = "proto3";

package bosdyn.api;

option java_outer_classname = "RobotCommandProto";

import "bosdyn/api/geometry.proto";
import "bosdyn/api/header.proto";
import "bosdyn/api/lease.proto";
import "bosdyn/api/trajectory.proto";
import "google/protobuf/any.proto";
import "google/protobuf/timestamp.proto";



/// A command for a robot to execute.
/// The server decides if a set of commands is valid for a given robot and configuration.
message RobotCommand {
    /// Commands which require control of entire robot.
    FullBodyCommand.Request full_body_command = 1;

    /// A mobility command for a robot to execute.
    MobilityCommand.Request mobility_command = 2;

}

/// Command specific feedback. Distance to goal, estimated time remaining, probability of
/// success, etc. Note that the feedback should directly mirror the command request.
message RobotCommandFeedback {
    /// Commands which require control of entire robot.
    FullBodyCommand.Feedback full_body_feedback = 2;

    /// Command to control mobility system of a robot.
    MobilityCommand.Feedback mobility_feedback = 1;

}

// Freeze all joints at their current positions (no balancing control)
message FreezeCommand {
    message Request {
        // Freeze command takes no args
    }

    message Feedback {
        /// Freeze command provides no feedback
    }
}

message FullBodyCommand {
    message Request {
        oneof command {
            StopCommand.Request stop_request = 1;
            FreezeCommand.Request freeze_request = 2;
            SelfRightCommand.Request selfright_request = 3;
            SafePowerOffCommand.Request safe_power_off_request = 4;

        }

        // Robot specific command parameters.
        google.protobuf.Any params = 100;
    }

    message Feedback {
        oneof feedback {
            StopCommand.Feedback stop_feedback = 1;
            FreezeCommand.Feedback freeze_feedback = 2;
            SelfRightCommand.Feedback selfright_feedback = 3;
            SafePowerOffCommand.Feedback safe_power_off_feedback = 4;

        }
    }
}

// Get robot a "ready" position
message SelfRightCommand {
    message Request {
        /// SelfRight command takes no args
    }

    message Feedback {
        /// SelfRight command provides no feedback
    }
}

/// Stop the robot in place with minimal motion.
message StopCommand {
    message Request {
        /// Stop command takes no args
    }

    message Feedback {
        /// Stop command provides no feedback
    }
}

/// Get robot into a position where it is safe to power down, then power down. If the robot has
/// fallen, it will power down directly. If the robot is not in a safe position, it will get to a
/// safe position before powering down. The robot will not power down until it is in a safe state.
message SafePowerOffCommand {
    message Request {
        // SafePowerOff command takes no args
    }

    message Feedback {
        enum Status {
            STATUS_UNKNOWN = 0;
            STATUS_POWER_ON = 1;
            STATUS_POWER_OFF = 2;
        }
        Status status = 1;
    }
}

message MobilityCommand {
    message Request {
        oneof command {
            SE2TrajectoryCommand.Request se2_trajectory_request = 1;
            SE2VelocityCommand.Request se2_velocity_request = 2;
            SitCommand.Request sit_request = 3;
            StandCommand.Request stand_request = 4;

        }

        // Robot specific command parameters.
        google.protobuf.Any params = 100;
    }

    message Feedback {
        oneof feedback {
            SE2TrajectoryCommand.Feedback se2_trajectory_feedback = 1;
            SE2VelocityCommand.Feedback se2_velocity_feedback = 2;
            SitCommand.Feedback sit_feedback = 3;
            StandCommand.Feedback stand_feedback = 4;

        }
    }
}

// Move along a trajectory in 2D space.
message SE2TrajectoryCommand {
    message Request {
        /// The timestamp (in robot time) by which a command must finish executing.
        /// This is a required field and used to prevent runaway commands.
        google.protobuf.Timestamp end_time = 1;

        SE2Trajectory trajectory = 2;
    }

    message Feedback {
        enum Status {
            STATUS_UNKNOWN = 0;

            /// The robot has arrived at the goal.
            STATUS_AT_GOAL = 1;

            /// The robot is attempting to go to a goal.
            STATUS_GOING_TO_GOAL = 2;
        }
        Status status = 1;
    }
}

message SE2VelocityCommand {
    message Request {
        /// The timestamp (in robot time) by which a command must finish executing. This is a
        /// required field and used to prevent runaway commands.
        google.protobuf.Timestamp end_time = 1;

        /// Desired planar velocity.
        SE2Velocity velocity = 2;

        /// Frame that velocity is specified in.
        Frame frame = 3;

        /// If set, limits how quickly velocity can change. Otherwise, robot may decide to limit.
        /// These values should be non-negative.
        SE2Velocity slew_rate_limit = 4;
    }

    message Feedback {
        /// Planar velocity commands provide no feedback
    }
}

message SitCommand {
    message Request {
        /// Sit commands take no args
    }

    message Feedback {
        /// Sit commands provide no feedback
    }
}

message StandCommand {
    message Request {
    }

    message Feedback {
        enum Status {
            STATUS_UNKNOWN = 0;
            STATUS_STANDING = 1;
            STATUS_NOT_STANDING = 2;
        }
        Status status = 1;
    }
}


message RobotCommandRequest {
    /// Common request header.
    RequestHeader header = 1;

    /// The Lease to show ownership of the robot.
    Lease lease = 2;

    // A command for a robot to execute. A command can be comprised of several subcommands.
    RobotCommand command = 3;

    /// Identifier provided by the time sync service to verify time sync between robot and client.
    string clock_identifier = 4;
}

message RobotCommandResponse {
    /// Common response header.
    ResponseHeader header = 1;

    /// Details about how the lease was used.
    LeaseUseResult lease_use_result = 2;

    enum Status {
        STATUS_UNKNOWN = 0;          ///< An unknown / unexpected error occurred.
        STATUS_OK = 1;               ///< Request was accepted.

        // Programming errors.
        STATUS_INVALID_REQUEST = 2;  ///< Request was invalid / malformed in some way.
        STATUS_UNSUPPORTED = 3;      ///< The robot does not understand this command.

        // Timesync errors.
        STATUS_NO_TIMESYNC = 4;      ///< Client has not done timesync with robot.
        STATUS_EXPIRED = 5;          ///< The command was received after its end_time had already passed.
        STATUS_TOO_DISTANT = 6;      ///< The command end time was too far in the future.

        // Robot hardware errors.
        STATUS_NOT_POWERED_ON = 7;   ///< The robot must be powered on to accept a command.
    }
    Status status = 3;

    /// Human-readable error description.  Not for programmatic analysis.
    string message = 4;

    /// Unique identifier for the command, If empty, command was not accepted.
    uint32 robot_command_id = 5;
}

message RobotCommandFeedbackRequest {
    /// Common request header.
    RequestHeader header = 1;

    /// Unique identifier for the command, provided by StartRequest.
    uint32 robot_command_id = 2;
}

message RobotCommandFeedbackResponse {
    /// Common response header.
    ResponseHeader header = 1;

    /// General status whether or not command is still processing.
    enum Status {
        /// Behavior execution is in an unknown / unexpected state.
        STATUS_UNKNOWN = 0;

        /// The robot is actively working on the command
        STATUS_PROCESSING = 1;

        /// The command was replaced by a new command
        STATUS_COMMAND_OVERRIDDEN = 2;

        /// The command expired
        STATUS_COMMAND_TIMED_OUT = 3;

        /// The robot is in an unsafe state, and will only respond to known safe commands.
        STATUS_ROBOT_FROZEN = 4;
    }
    Status status = 2;

    /// Human-readable status message.  Not for programmatic analysis.
    string message = 3;

    /// Command specific feedback.
    RobotCommandFeedback feedback = 4;
}

message ClearBehaviorFaultRequest {
    /// Common request header.
    RequestHeader header = 1;

    /// The Lease to show ownership of the robot.
    Lease lease = 2;

    /// Unique identifier for the error
    uint32 behavior_fault_id = 3;
}

message ClearBehaviorFaultResponse {
    /// Common response header.
    ResponseHeader header = 1;

    /// Details about how the lease was used.
    LeaseUseResult lease_use_result = 2;

    enum Status {
        /// An unknown / unexpected error occurred.
        STATUS_UNKNOWN = 0;

        /// The BehaviorFault has been cleared.
        STATUS_CLEARED = 1;

        /// The BehaviorFault could not be cleared.
        STATUS_NOT_CLEARED = 2;
    }
    Status status = 3;
}
